File: Sources/ATTNCustomEvent.m
//
//  ATTNCustomEvent.m
//  attentive-ios-sdk
//
//  Created by Wyatt Davis on 3/20/23.
//

#import "ATTNCustomEvent.h"
#import "attentive_ios_sdk_framework/attentive_ios_sdk_framework-Swift.h"

@implementation ATTNCustomEvent

- (instancetype)initWithType:(NSString *)type properties:(NSDictionary<NSString *, NSString *> *)properties {
  if (self = [super init]) {
    [ATTNParameterValidation verifyNotNil:type inputName:@"type"];
    [ATTNParameterValidation verifyStringOrNil:type inputName:@"type"];

    NSString *invalidCharInType = [ATTNCustomEvent findInvalidCharacterInType:type];
    if (invalidCharInType != nil) {
      NSLog(@"Invalid character '%@' in CustomEvent type '%@'", invalidCharInType, type);
      return nil;
    }

    [ATTNParameterValidation verifyNotNil:properties inputName:@"properties"];
    [ATTNParameterValidation verify1DStringDictionaryOrNil:properties inputName:@"properties"];

    for (NSString *key in properties.allKeys) {
      NSString *invalidCharInKey = [ATTNCustomEvent findInvalidCharacterInPropertiesKey:key];
      if (invalidCharInKey != nil) {
        NSLog(@"ERROR: Invalid character '%@' in CustomEvent property key '%@'", invalidCharInKey, key);
        return nil;
      }
    }

    _type = type;
    _properties = [[NSDictionary alloc] initWithDictionary:properties];
  }

  return self;
}

+ (NSString *)findInvalidCharacterInType:(NSString *)type {
  NSArray *invalidCharacters = @[ @"\"", @"'", @"(", @")", @"{", @"}", @"[", @"]", @"\\", @"|", @"," ];
  return [self findCharacterInString:type charactersToFind:invalidCharacters];
}

+ (NSString *)findInvalidCharacterInPropertiesKey:(NSString *)key {
  NSArray *invalidCharacters = @[ @"\"", @"{", @"}", @"[", @"]", @"\\", @"|" ];

  return [self findCharacterInString:key charactersToFind:invalidCharacters];
}

+ (NSString *)findCharacterInString:(NSString *)input charactersToFind:(NSArray *)charactersToFind {
  for (NSString *character in charactersToFind) {
    if ([input rangeOfString:character].location != NSNotFound) {
      return character;
    }
  }
  return nil;
}

@end

File: Sources/Migration/Public/Events/ATTNCustomEvent.swift
//
//  ATTNCustomEvent.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-30.
//

import Foundation

@objc(ATTNCustomEvent)
public final class ATTNCustomEvent: NSObject, ATTNEvent {
  @objc public let type: String
  @objc public let properties: [String: String]

  @objc(initWithType:properties:)
  public init?(type: String, properties: [String: String]) {
    ATTNParameterValidation.verifyNotNil(type as NSObject, inputName: "type")
    ATTNParameterValidation.verifyStringOrNil(type as NSObject, inputName: "type")
    if let invalidCharInType = ATTNCustomEvent.findInvalidCharacter(in: type) {
      print("Invalid character \(invalidCharInType) in CustomEvent type \(type)")
      return nil
    }

    ATTNParameterValidation.verifyNotNil(properties as NSObject, inputName: "properties")
    ATTNParameterValidation.verify1DStringDictionaryOrNil(properties as NSDictionary, inputName: "properties")
    if let _ = properties.keys
      .map({ ATTNCustomEvent.findInvalidCharacterInPropertiesKey(key: $0) })
      .first(where: { $0 != nil }) {
      return nil
    }

    self.type = type
    self.properties = properties
  }
}

fileprivate extension ATTNCustomEvent {
  static func findInvalidCharacter(in type: String) -> String? {
    let invalidCharacters = ["\"", "'", "(", ")", "{", "}", "[", "]", "\\", "|", ","]
    return findCharacter(in: type, charactersToFind: invalidCharacters)
  }

  static func findInvalidCharacterInPropertiesKey(key: String) -> String? {
    let invalidCharacters = ["\"", "{", "}", "[", "]", "\\", "|"]
    return findCharacter(in: key, charactersToFind: invalidCharacters)
  }

  static func findCharacter(in input: String, charactersToFind: [String]) -> String? {
    for character in charactersToFind {
      if input.range(of: character) != nil {
        return character
      }
    }
    return nil
  }
}
